Fix: IPAMBlock GC breaks flannel migration

squ94wk opened the pull request:
## Description

This fixes a regression in the kube-controllers ipam code that breaks the flannel migration.

The empty block garbage collection breaks the flannel migration, because it cleans up blocks with IPs that are in use only by unmigrated Flannel nodes. This deletes the routes on Calico nodes to IPs in those blocks.

The kube-controllers already sometimes skips releasing blocks, e.g. when it's the node's only affined block.
I just add another check that gets the node currently affined to the block and check if it has labels that exist only on nodes that haven't been migrated yet.

I have successfully tested the migration on a cluster running 1.21.5 and flannel 0.11.0.
I believe there aren't any automated tests for the migration. Otherwise the regression would've been caught in the first place.

The new code needs access to the labels the migration controller uses. I chose to export that variable in favour of copy-pasting. That's a personal choice. I'm fine with either one. One could also have an exported var and a non-exported var. That would reduce the diff of this MR, but also be "dirty".

## Related issues/PRs

fixes https://github.com/projectcalico/calico/issues/5349

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
Fix garbage collection of empty blocks during flannel migration
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/projectcalico/calico?pullRequest=5459) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/projectcalico/calico?pullRequest=5459) before we can accept your contribution.<br/><hr/>**jmende** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/projectcalico/calico?pullRequest=5459) it.</sub>

caseydavenport commented on kube-controllers/pkg/controllers/node/ipam.go:
```diff
@@ -568,6 +571,26 @@ func (c *ipamController) checkEmptyBlocks() error {
 		}
 
+		// Get node to inspect labels
+		node, err := c.client.Nodes().Get(context.TODO(), nodeName, options.GetOptions{})
```
We can't do a Node.Get() on every iteration here - for large clusters that's going to be a substantial increase in work performed by this controller and load on the API server.

 I believe we already have a watch configured that is caching node objects - we should be able to get the Node object from that cache instead.

caseydavenport replied:
Here's where we created the IndexerInformer: https://github.com/projectcalico/calico/blob/3ca528bed2d9af244d70ca7cb581d0af107e148b/kube-controllers/pkg/controllers/node/controller.go#L54

You should be able pass it to the IPAM controller as a new argument, something like this:

```go
NewIPAMController(cfg, calicoClient, k8sClientset, nc.nodeInformer.Indexer())
```

Then, instead of this `Get()` command here, you can call the indexer's `GetByKey(<node name>)` command to return the cached node value.

squ94wk replied:
Thanks for the input. I changed it as you suggested.

caseydavenport commented on kube-controllers/pkg/controllers/node/ipam.go:
```diff
@@ -568,6 +571,26 @@ func (c *ipamController) checkEmptyBlocks() error {
+
+		// We cannot release blocks affined to nodes undergoing migration from Flannel
+		// Its Pod's IPs are not managed by Calico
+		for labelName, labelVal := range node.ObjectMeta.Labels {
```
Would be nice to see this encapsulated in a helper function - I think it simplifies the loop continuation logic as well.

e.g.,

```go
if nodeIsBeingMigrated(node) {
  logc.Info("node is currently undergoing a migration from flannel, do not release blocks")
  continue
}
```

caseydavenport replied:
If we attach the helper to the ipam controller, it can even include the Get() call above.

caseydavenport reviewed (changes requested):
Thanks for the PR! I've added some thoughts as comments. Let me know what you think.

caseydavenport commented on kube-controllers/pkg/controllers/node/ipam.go:
```diff
@@ -568,16 +574,26 @@ func (c *ipamController) checkEmptyBlocks() error {
 		// Find the actual block object.
-		block, ok := c.allBlocks[blockCIDR]
-		if !ok {
+		block, migrating := c.allBlocks[blockCIDR]
```
Is this an unintended change? I think this should go back to being `block, ok := c.allBlocks[blockCIDR]`

squ94wk replied:
yes, updated it.

caseydavenport commented on kube-controllers/pkg/controllers/node/ipam.go:
```diff
@@ -568,16 +574,26 @@ func (c *ipamController) checkEmptyBlocks() error {
+			continue
+		}
+		if migrating {
+			logc.Info("Node affined to block is currently undergoing a migration from Flannel, skipping affinity release")
```
I believe you're missing a `continue` in this block

squ94wk replied:
thanks.

squ94wk replied:
I haven't tested the migration again.
I can probably test it tomorrow however.
Do you have tests for the migration yourself?

caseydavenport commented on kube-controllers/pkg/controllers/node/ipam.go:
```diff
@@ -568,16 +574,27 @@ func (c *ipamController) checkEmptyBlocks() error {
+		// During a Flannel migration, we can only migrate blocks affined to nodes that have already undergone the migration
+		migrating, err := c.nodeIsBeingMigrated(node)
+		if err != nil {
+			logc.Warn("Couldn't find node affined to empty block, skipping affinity release")
```
I don't think we can skip in this case - there are valid scenarios where blocks have been orphaned and the node no longer exists, and we need the garbage collector to properly release those blocks. We can only skip if the node exists and is actively undergoing a flannel migration.

squ94wk replied:
Then the helper should catch if the node doesn't exist and return that as well.
In case of a different error, we should assume that we just couldn't get the node from the API.?

caseydavenport replied:
Yeah, I think it's OK to skip on other errors - those are true "error" scenarios. But NotFound shouldn't be treated as an error :+1:

caseydavenport replied:
Sorry I missed this before - this log should include a `WithError` so we log out the original error.

e.g.,

```
logc.WithError(err).Warn("Failed to check if node is undergoing flannel migration")
```

caseydavenport commented on kube-controllers/pkg/controllers/node/ipam.go:
```diff
@@ -967,6 +984,37 @@ func (c *ipamController) nodeExists(knode string) bool {
+		}
+		return false, fmt.Errorf("failed to check node for migration status: %w", err)
+	}
+	if !ok {
```
I believe this is the `IsNotFound` case - we probably want to structure this like so:

```go
obj, ok, err := c.nodeIndexer.GetByKey(name)
if !ok {
  // Node doesn't exist, so isn't being migrated.
  return false, nil
}
if err != nil {
  return false, err
}
...

caseydavenport reviewed (commented):
Looking pretty close!

The last thing I think we need is a test for this case in the IPAM GC tests. You can probably start with this test as a good example: https://github.com/projectcalico/calico/blob/4d17273be284eeef8f0c954e17ef1d477f2277ab/kube-controllers/pkg/controllers/node/ipam_test.go#L956

squ94wk commented:
In `ipam_test.go`, the constructor for the Controller now needs a (fake) node indexer.
It seems like this doesn't exist yet. Should I write a fake_nodeIndexer.go using a fakeNodeClient to implement the member functions?

squ94wk commented:
When I create a nodeIndexer in the test like so (like in main.go):
```
factory := informers.NewSharedInformerFactory(cs, 0)
nodeInformer := factory.Core().V1().Nodes().Informer()
ni = nodeInformer.GetIndexer()
```
and pass it in the constructor, after I create a node with the fake client, the informer doesn't have it.
I'm not too familiar with the informers, indexers.
Do you have an idea, why it doesn't work?

caseydavenport commented:
Ah, yes. I recall hitting this before.

The fake clientset by default doesn't hook into the informer properly. It requries some manual hooking up to make it work as expected. I recall writing that code once but it was a long time ago and I can't seem to find it any more....

One thing you might do to make this easier is to instead pass a `cache.Store` to the controller instead. Then, it's easy enough to build a `MockStore` object that implements the [interface](https://pkg.go.dev/k8s.io/client-go/tools/cache#Store) and pass that.


For example:

```go
type fakeNodeStore {
  cs clientSet
}

func NewFakeNodeStore(cs clientSet) { . . .}

. . . 

func (f *fakeStore) GetByKey(k string) {
  return cs.Nodes().Get(key)
}
```

I think that would give you what you need for this? In the meantime I'll keep looking for a better approach, but I think that this would work well enough.

squ94wk commented:
I added a fakeNodeIndexer. I left all other member functions unimplemented for now.
Also, do you have any rules on comments?
I saw the fake calico client had comments on most exported functions, but not all.

I also added a lookup for the Kubernetes node belonging to the Calico node. I didn't know there was another indirection.
The informer thing still doesn't work without the fakeNodeIndexer.

caseydavenport commented:
/sem-approve

caseydavenport commented:
@squ94wk how about for now we revert the test changes and we can get this merged so long as existing tests pass - I don't want to hold this up and let it rot. 

We can work on tests in a separate follow-on PR.

squ94wk commented:
Fine with me.
I rebased on master and pushed HEAD~2. This removed the commit with the fakeNodeIndexer and the added test.

caseydavenport commented:
/sem-approve

caseydavenport commented:
@squ94wk think you're failing a code linter:

```
../lib.Makefile:541: recipe for target 'check-fmt' failed
```

Should be able to run `make fix` to clean it up, and `make check-fmt` to verify.

caseydavenport commented:
/sem-approve

caseydavenport commented:
> errors in package: [/go/src/github.com/projectcalico/calico/kube-controllers/pkg/controllers/node/ipam_test.go:110:37: too few arguments in call to NewIPAMController]"

Hitting some build errors

caseydavenport commented:
/sem-approve

caseydavenport commented:
@squ94wk nice, tests passed and merged! 

Think we still need a follow-on PR to add a specific test for this scenario though.
